velox/common/memory/tests/SharedArbitratorTest.cpp

```diff
@@ -426,6 +433,69 @@ DEBUG_ONLY_TEST_F(SharedArbitrationTest, reclaimFromOrderBy) {
   }
 }
 
+DEBUG_ONLY_TEST_F(
+    SharedArbitrationTest,
+    filterProjectInitTriggeredArbitration) {
+  SCOPED_TESTVALUE_SET(
+      "facebook::velox::memory::{}::FakeMemoryOperator::FakeMemoryOperator",
+      std::function<void(std::pair<MemoryPool*, memory::Allocation*>*)>(
+          [&](std::pair<MemoryPool*, memory::Allocation*>* allocationHook) {
+            allocationHook->first->allocateNonContiguous(
+                AllocationTraits::numPages(1 * MB), *(allocationHook->second));
+          }));
+  setupMemory(2.5 * MB, 0);
+  auto rootPool = memoryManager_->addRootPool("root-0", 1 * MB);
```

Comment:
I think you could simplify the test case by creating two tasks:
(1) one non-reclaimable fake task;
(2) project filter task
You allocate all memory to fake task (block is through inject callback provided by the fake memory operator) and then start the project filter task and expect project filter task fails with oom instead of hanging.

Reply:
I actually tried this way before I think. It took me some time to figure out the current test. To reproduce a hang with previous code, this will probably not work because when project filter task attempts to allocate memory, it goes into arbitration but itself will not be part of arbitration as its current memory usage is 0. So in order to trigger itself to be participating in the arbitration process, it needs to have at least some memory allocated.

Reply:
I think if the two tasks share the same root pool or query ctx. It should work?

Reply:
Ok. It won't work with the task order things.

Reply:
Ok. The following test procedure might work?
```
T0 set testvalue in createDriversLocked to block the first task driver creation and wait for unblock.
T1 create a shared query ctx -> set up memory arbitrator with zero initial capacity
T2 in a test thread to create filter project task to execute and expect success
T3 create fake memory operator task and setup allocationCb_ to inject allocation once (allocateBytes is sufficient) to allocate all the memory. Fake memory operator is reclaimable. After allocation succeed, unblock filter project start in T2
```